Improved Bazin Fit

sidchaini opened the pull request:
Johann and I were looking at the Bazin fit and noticed that the Bazin fit was failing for PLAsTiCC objects unexpectedly. For example, the fit was failing for a microlensing example (object_id: 130779836), which is an example of a µ-lens from a single lens. Here's the failed fit:
![before](https://user-images.githubusercontent.com/40721514/120375430-6d0bb200-c338-11eb-935a-ccec2981d1e7.png)

The ratio of `tfall/trise` is supposed to be `>1`, as mentioned in Section 3, [Dai et. al. (2017)](https://arxiv.org/pdf/1701.05689.pdf).

Thus, Johann and I made changes to the 3 functions: `bazin()`, `errfunc()` and `fit_scipy()` coded up the following changes:
1.  Bazin is modified to take the ratio `r=tfall/trise` instead of `trise`. This allows us to enforce the condition `r>1` on the fit.
2.  Flux_err is used as a parameter to errfunc: `abs(flux - bazin(time, *params)) / fluxerr`
3.  Bounds can be added to `fit_scipy()` and a better original guess was chosen to again enforce the condition `r>1`.

Here's the same object as above (object_id: 130779836), now fit with the new code:
![after](https://user-images.githubusercontent.com/40721514/120376841-fa033b00-c339-11eb-9c4b-843fc2e0ce40.png)

There is a clear improvement in the fit for the "u", "r" and "z" bands.

sidchaini committed:
modified bazin, errfunc and fit_scipy

Modified the functions (bazin, errfunc and fit_scipy) written by Johann. These consider fluxerr as a parameter.

sidchaini committed:
modified bazin, errfunc and fit_scipy

Modified the functions (bazin, errfunc and fit_scipy) written by Johann. These consider fluxerr as a parameter.

sidchaini committed:
modified bazin, errfunc and fit_scipy

Modified the functions (bazin, errfunc and fit_scipy) written by Johann. These consider fluxerr as a parameter.

sidchaini committed:
modified bazin, errfunc and fit_scipy

Modified the functions (bazin, errfunc and fit_scipy) written by Johann. These consider fluxerr as a parameter.

sidchaini committed:
modified bazin, errfunc and fit_scipy

Modified the functions (bazin, errfunc and fit_scipy) written by Johann. The functions errfunc and fit_scipy consider fluxerr as a parameter, while bazin now takes r (=tfall/trise) as a parameter instead of trise. r>1 is enforced.

sidchaini committed:
modified bazin, errfunc and fit_scipy

Modified the functions (bazin, errfunc and fit_scipy) written by Johann. The functions errfunc and fit_scipy consider fluxerr as a parameter, while bazin now takes r (=tfall/trise) as a parameter instead of trise. r>1 is enforced.

sidchaini committed:
Add files via upload

sidchaini committed:
Fixed typo

Fixed typo in build_plasticc_canonical

sidchaini committed:
modified bazin, errfunc and fit_scipy

Modified the functions (bazin, errfunc and fit_scipy) written by Johann. The functions errfunc and fit_scipy consider fluxerr as a parameter, while bazin now takes r (=tfall/trise) as a parameter instead of trise. r>1 is enforced.

johannct commented on resspect/tests/test_bazin.py:
```diff
@@ -71,8 +76,9 @@ def test_fit_scipy():
     
     time = data['mjd'].values
     flux = data['flux'].values
+    fluxerr = data['fluxerr'].values
```
change to fluxerr = 1 to fit in the unit test as it was done previously.

johannct replied:
this should solve the current unit test failure

sidchaini replied:
Okay, done. That solved it, thanks!

sidchaini committed:
changed fluxerr=1 for unit test

johannct reviewed (approved):
LGTM, but I would like @emilleishida to have a look, as ultimately it has to fit the global picture.

emilleishida commented:
This is great, thanks @siddharthchaini and @johannct !
I was just trying to see how we can reconcile this with other parts of the code.

Would it be possible to make the Bazin funcion still receive `trise` and `tfall` and make the ratio conversion inside the function?

this would keep the same column names and avoid the need of further changes in other parts of the code. 
if  I change the `database.py` entries to comply with the column names it will not be compatible when reading previously fitted files.

johannct commented:
@emilleishida the function needs to take the parameter to be fitted as argument, and the parameter to be fitted needs to be the ratio if one wants to set the lower boundary in the fitter..... What can be done is to decompose r into trise and tfall *after* the fit.

emilleishida commented:
thanks @johannct , yes I think that would do. 
my main concern is to keep the column names in the output files

sidchaini committed:
fixed bug in fit_scipy

fixed a bug where the variable "t" was used instead of "time", always leading to a single condition from the except block.

sidchaini committed:
Revert "fixed bug in fit_scipy"

This reverts commit acf328050aee924500dfde644fd138c661b2e0ec.

sidchaini committed:
Revert "Add files via upload"

This reverts commit 154ffe37642da66b8200a81df799b339a1683d48.

sidchaini committed:
Revert "modified bazin, errfunc and fit_scipy"

This reverts commit 9246e0085383f46963f01b371aa79a50f0f83a78.

sidchaini committed:
Revert "modified bazin, errfunc and fit_scipy"

This reverts commit 5b7cdec9278f0e85ae514e2fc7ed2fcaaa4ef163.

sidchaini committed:
fixed bazin and reverted test_bazin

Fixed bazin.py to output trise instead of r. Also, reverted test_bazin to original.

sidchaini committed:
Update test_bazin.py

Fixed typo leading to bug

sidchaini committed:
Fixed typo

Replaced the variable "t" with "time"

sidchaini committed:
Update bazin.py

sidchaini committed:
Fixed typo trsise -> trise

Fixed typo. Replaced "trsise" to "trise"

siddharthchaini commented:
@emilleishida @johannct I think the latest commit should have implemented all of the above changes. I also added some extra lines to fit_scipy() to make it more readable. Let me know if anything else is required.

emilleishida commented:
hi @siddharthchaini apologies for the super late reply on this. 
I was trying your modifications and I got an error trying to this light curve from SNPCC: `DES_SN813144.DAT`

Can you please try to fit this light curve and see if you get the same error?

`---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/tmp/ipykernel_207281/4206856705.py in <module>
----> 1 lc.fit_bazin_all()

/media/emille/git/COIN/RESSPECT_work/siddharth/venv/resspect_siddarth/lib/python3.7/site-packages/resspect-0.1-py3.7.egg/resspect/fit_lightcurves.py in fit_bazin_all(self)
    511         default_bazin_features = ['None'] * len(self.bazin_features_names)
    512         for each_band in self.filters:
--> 513             best_fit = self.fit_bazin(each_band)
    514             if (best_fit.size > 0) and (not np.isnan(np.sum(best_fit))):
    515                 self.bazin_features.extend(best_fit.tolist())

/media/emille/git/COIN/RESSPECT_work/siddharth/venv/resspect_siddarth/lib/python3.7/site-packages/resspect-0.1-py3.7.egg/resspect/fit_lightcurves.py in fit_bazin(self, band)
    466 
    467         # fit Bazin function
--> 468         bazin_param = fit_scipy(time - time[0], flux, fluxerr)
    469         return bazin_param
    470 

/media/emille/git/COIN/RESSPECT_work/siddharth/venv/resspect_siddarth/lib/python3.7/site-packages/resspect-0.1-py3.7.egg/resspect/bazin.py in fit_scipy(time, flux, fluxerr)
    157               [a_bounds[1], b_bounds[1], t0_bounds[1], tfall_bounds[1], r_bounds[1]]]
    158 
--> 159     result = least_squares(errfunc, guess, args=(time, flux, fluxerr), method='trf', loss='linear',bounds=bounds)
    160 
    161     a_fit,b_fit,t0_fit,tfall_fit,r_fit = result.x

/media/emille/git/COIN/RESSPECT_work/siddharth/venv/resspect_siddarth/lib/python3.7/site-packages/scipy/optimize/_lsq/least_squares.py in least_squares(fun, x0, jac, bounds, method, ftol, xtol, gtol, x_scale, loss, f_scale, diff_step, tr_solver, tr_options, jac_sparsity, max_nfev, verbose, args, kwargs)
    806 
    807     if not in_bounds(x0, lb, ub):
--> 808         raise ValueError("`x0` is infeasible.")
    809 
    810     x_scale = check_x_scale(x_scale, x0)

ValueError: `x0` is infeasible.`

sidchaini committed:
Update README.md

removed extra whitespace from code

sidchaini committed:
Update bazin.py

Added clipping to make sure guesses are within bounds

siddharthchaini commented:
@emilleishida I was able to reproduce the error - it was taking place because the initial guess was out of bounds. 

I have fixed it by clipping the guesses so that they stay within bound. Could you check and see if there are any other bugs?

emilleishida commented:
hi @siddharthchaini this works fine, thanks!

just a note for future reference, in `bazin.py` file, lines 132-137 it is best to avoid `np.inf` and use a very large number instead. 
Otherwise there are some errors that appear down the line. 

I will approve this and make the necessary changes to the main branch. 
thank you deeply for the work on this!
